src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb

```diff
@@ -49,19 +49,22 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.CodeModel
                 ' If tests are written that require multiple projects, additional support will need to be added.
                 Dim project = workspace.CurrentSolution.Projects.Single()
 
+                Dim threadingContext = workspace.ExportProvider.GetExportedValue(Of IThreadingContext)
+
                 Dim state = New CodeModelState(
-                    workspace.ExportProvider.GetExportedValue(Of IThreadingContext),
+                    threadingContext,
                     mockServiceProvider,
                     project.LanguageServices,
                     mockVisualStudioWorkspace,
-                    Nothing)
+                    New ProjectCodeModelFactory(mockVisualStudioWorkspace, mockServiceProvider, threadingContext))
 
-                Dim mockTextManagerAdapter = New MockTextManagerAdapter()
+                Dim projectCodeModel = DirectCast(state.ProjectCodeModelFactory.CreateProjectCodeModel(project.Id, Nothing), ProjectCodeModel)
 
-                For Each documentId In project.DocumentIds
+                For Each document In project.Documents
                     ' Note that a parent is not specified below. In Visual Studio, this would normally be an EnvDTE.Project instance.
-                    Dim fcm = FileCodeModel.Create(state, parent:=Nothing, documentId:=documentId, textManagerAdapter:=mockTextManagerAdapter)
-                    mockVisualStudioWorkspace.SetFileCodeModel(documentId, fcm)
+                    Dim fcm = projectCodeModel.GetOrCreateFileCodeModel(document.FilePath, parent:=Nothing)
+                    fcm.Object.TextManagerAdapter = New MockTextManagerAdapter()
```

Comment:
I will totally admit assigning this here is hacky, but it's the best I can do without a much larger refactoring if we want to get this in for Preview 2.

Reply:
That's fine. I was equally conflicted as I read through the rest of the code, but hadn't said anything yet. Glad we're of the same mind. :smile:

Reply:
I *did* try to thread it through and ended up in a much bigger mess...

Reply:
Can we at least change it to the `Testaccessor` pattern, and make the setter of the main property `private`?

Reply:
Given this is being considered for Preview 2, I'm not going to touch this _right now_. I think that's a great suggestion and if it merges in this form I'll dig into this further.

The other approach I think makes more sense is just exposing this either as a MEF or workspace service or IServiceProvider service, or something else. The core operation needed is "give me a EnvDTE.TextPoint" which in VS has to be done through the shell. For tests, we just use a mock. Any of our standard dependency injection patterns would then mean we avoid the need of this having to do "special" stuff in this way.

Reply:
And oh: if this is rejected for Preview 2, then I'll give another crack at this for Preview 3.

Reply:
...and of course as I write that, it's approved for Preview 2. :smile: